### PR TITLE
Fix 13010 - Ensure pytest.approx in sequence/collections only process instances of Number

### PR DESCRIPTION
**Pull request by marcelotrevisani**
#13010 
Add a check to ensure that `approx` calculations only process instances of `Number`, preventing errors with incompatible types.

### CODE REVIEW

File: testing/python/approx.py [hunk 1]
```diff
@@ -377,6 +377,22 @@ def test_bool(self):
...
+                "",
+                r"  comparison failed. Mismatched elements: 2 / 3:",
+                rf"  Max absolute difference: {SOME_FLOAT}",
+                rf"  Max relative difference: {SOME_FLOAT}",
+                r"  Index \| Obtained\s+\| Expected\s+",
+                r"\s*0\s*\|\s*1\.1\s*\|\s*1\.0\s*±\s*1\.0e\-06\s*",
+                r"\s*2\s*\|\s*word\s*\|\s*different\s*",
+            ],
+            verbosity_level=2,
+        )
```
Comment (nicoddemus) [inline testing/python/approx.py:394]: While at it, let's add an assert that works:

```suggestion
        )
        assert [1.1, 2, "word"] == pytest.approx([1.1, 2, "word"])
```

File: testing/python/approx.py [hunk 2]
```diff
@@ -377,6 +377,22 @@ def test_bool(self):
 
         assert err.match(r"approx\(\) is not supported in a boolean context")
 
+    def test_list_with_str(self, assert_approx_raises_regex):
```
Comment (nicoddemus) [inline testing/python/approx.py:380]:
```suggestion
    def test_list_with_str(self, assert_approx_raises_regex) -> None:
        """approx should work on sequences that also contain non-numbers (#13010)."""
```

**Review by nicoddemus — APPROVED** [review]
Thanks @marcelotrevisani for the PR, appreciated!

Left some small comments, other than that, LGTM.

File: changelog/13010.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Add a check to ensure that :func:`pytest.approx` calculations only process instances of ``Number``, preventing errors with incompatible types.
```
Comment (nicoddemus) [inline changelog/13010.bugfix.rst:1]:
```suggestion
:func:`pytest.approx` now can compare collections that contain numbers and non-numbers mixed.
```

Also please rename this file to `13010.improvement.rst`. :+1:

**Comment by marcelotrevisani** [comment]
all done @nicoddemus !
thanks for taking the time to review this, appreciated! :)

File: src/_pytest/python_api.py [hunk 4]
```diff
@@ -340,14 +341,14 @@ def _repr_compare(self, other_side: Sequence[float]) -> list[str]:
...
         ):
             if approx_value != other_value:
-                abs_diff = abs(approx_value.expected - other_value)
-                max_abs_diff = max(max_abs_diff, abs_diff)
-                if other_value == 0.0:
-                    max_rel_diff = math.inf
-                else:
-                    max_rel_diff = max(max_rel_diff, abs_diff / abs(other_value))
+                if isinstance(approx_value.expected, Number):
+                    abs_diff = abs(approx_value.expected - other_value)
```
Comment (marcelotrevisani) [inline src/_pytest/python_api.py:345]: mypy is complaining in this part
``Unsupported operand types for - ("Number" and "float")``

should I disregards this or better to change to 
```python
if isinstance(approx_value.expected, (Decimal, Complex)):
```
?

@nicoddemus

Comment (nicoddemus) [reply src/_pytest/python_api.py:345]: Let's use `(Complex, Decimal)` as we do in other parts of the code. 👍

**Comment by marcelotrevisani** [comment]
I sorted all comments and problems. If you have new suggestions please let me know. Thanks!

**Comment by nicoddemus** [comment]
FYI changed to a try/except instead of an instance check because mypy was correctly complaining about `Decimal` and `float` -- this is safer backward-compatibility wise too.

**Comment by marcelotrevisani** [comment]
> FYI changed to a try/except instead of an instance check because mypy was correctly complaining about `Decimal` and `float` -- this is safer backward-compatibility wise too.

Indeed, I believe your modification is better, thanks!

**Pull request merged**